pkg/gateway/server/dispatcher/dispatcher.go

```diff
@@ -30,129 +31,196 @@ type Dispatcher struct {
-	key := namespace + "/" + modelProviderName
-	// Check the map with the read lock.
-	d.modelLock.RLock()
-	u, ok := d.modelUrls[key]
-	d.modelLock.RUnlock()
-	if ok && (u.Hostname() != "127.0.0.1" || engine.IsDaemonRunning(u.String())) {
-		if u.Host == "api.openai.com" {
-			return u, d.openAICred, nil
+func (d *Dispatcher) startProvider(ctx context.Context, providerType types.ToolReferenceType, namespace, providerName string, extraEnv ...string) (url.URL, error) {
+	thread := &v1.Thread{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      system.ThreadPrefix + providerName,
+			Namespace: namespace,
+		},
+		Spec: v1.ThreadSpec{
+			SystemTask: true,
+		},
+	}
+
+	if err := d.client.Get(ctx, kclient.ObjectKey{Namespace: thread.Namespace, Name: thread.Name}, thread); apierrors.IsNotFound(err) {
+		if err = d.client.Create(ctx, thread); err != nil {
+			return url.URL{}, fmt.Errorf("failed to create thread: %w", err)
 		}
-		return u, "", nil
+	} else if err != nil {
+		return url.URL{}, fmt.Errorf("failed to get thread: %w", err)
 	}
 
-	d.modelLock.Lock()
-	defer d.modelLock.Unlock()
+	var providerToolRef v1.ToolReference
+	if err := d.client.Get(ctx, kclient.ObjectKey{Namespace: namespace, Name: providerName}, &providerToolRef); err != nil || providerToolRef.Spec.Type != providerType {
+		return url.URL{}, fmt.Errorf("failed to get model provider: %w", err)
+	}
 
-	// If we didn't find anything with the read lock, check with the write lock.
-	// It could be that another thread beat us to the write lock and added the model provider we desire.
-	u, ok = d.modelUrls[key]
-	if ok && (u.Hostname() != "127.0.0.1" || engine.IsDaemonRunning(u.String())) {
-		if u.Host == "api.openai.com" {
-			return u, d.openAICred, nil
+	credCtx := []string{string(providerToolRef.UID), providerTypeToGenericCredContext[providerType]}
+	if providerToolRef.Status.Tool == nil {
+		return url.URL{}, fmt.Errorf("provider tool reference %q has not been resolved", providerName)
+	}
+
+	// Ensure that the provider has been configured so that we don't get stuck waiting on a prompt.
+	mps, err := providers.ConvertProviderToolRef(providerToolRef, nil)
+	if err != nil {
+		return url.URL{}, fmt.Errorf("failed to convert model provider: %w", err)
+	}
+	if len(mps.RequiredConfigurationParameters) > 0 {
+		cred, err := d.gptscript.RevealCredential(ctx, credCtx, providerName)
+		if err != nil {
+			return url.URL{}, fmt.Errorf("model provider is not configured: %w", err)
+		}
+
+		mps, err = providers.ConvertProviderToolRef(providerToolRef, cred.Env)
+		if err != nil {
+			return url.URL{}, fmt.Errorf("failed to convert model provider: %w", err)
+		}
+
+		if len(mps.MissingConfigurationParameters) > 0 {
+			return url.URL{}, fmt.Errorf("model provider is not configured: missing configuration parameters %s", strings.Join(mps.MissingConfigurationParameters, ", "))
```

Comment:
```suggestion
			return url.URL{}, fmt.Errorf("provider is not configured: missing configuration parameters %s", strings.Join(mps.MissingConfigurationParameters, ", "))
```

Reply:
Done. Found a few more in this file, too.